### PR TITLE
Added TypeSript definition

### PR DESCRIPTION
**Pull request by T4cC0re**
Hey, i'd love if you could merge this and create a new version on npm so all TypeScript users can use your awesome library with a definiton.

Thanks!

### CODE REVIEW

**Comment by butterkekstorte** [comment]
Oh, nevermind. This causes the using typescript to not compile successfully. I'll fix it and report back

**Comment by butterkekstorte** [comment]
Was just a mistake in the import statement (readme.md). Defnintion itself was fine.

**Comment by jessetane** [comment]
I don't know anything about typescript, so I can't really review this. I am ok with adding it since it doesn't appear to hurt anything, though I'd prefer you didn't add the line to the docs - if typescript ppl want to use this and can do so unobtrusively that's fine but I think presentation should remain as a JavaScript library.

**Comment by butterkekstorte** [comment]
Ok, i'll remove that from the readme, also i found a little bug in the definition, so I would like to fix that first

**Comment by jessetane** [comment]
Hm, does it not make you nervous that you don't have a way to test this?

**Comment by jessetane** [comment]
iirc someone else already added an .npmignore that ignores the tests, so you shouldn't have to worry about install size. if it's possible to somehow add a test for what you're adding it might be a good idea.

**Comment by butterkekstorte** [comment]
Sure, i'll look into it.
But then i will probably need to add typescript to the devDependencies and a seperate travis-ci test.
As a Typescript definition is not really teastable as a js file, but rather just seeing if it compiles.

**Comment by jessetane** [comment]
not sure what to say... does it make more sense to just fork and have a queue-ts package with a readme that links back here? if it's possible to introduce a broken typescript definition does it not also make sense to test it?

**Comment by butterkekstorte** [comment]
Okay, i have removed the Typescript how-to and moved it into the definition itself (for those who are using typescript) So it won't throw off regular JS users.
Also added a compilation test for the TypeScript definition.

Is this ok for you?

**Comment by jessetane** [comment]
Yeah, I mean I will merge this if you want. However, I'm curious about your thoughts on whether this is the best way of solving this problem - for example, what happens when someone else comes along decides they need to add X for compile-to dialect Y? Considering the amount of stuff that needs to get added, is it possible that it makes more sense to create a wrapper package with this library as a dependency instead?

Like I say, if you feel strongly this is the best way I will merge, just wondering if we can't think of anything more elegant.

**Comment by butterkekstorte** [comment]
Another way could be to have the definitions externally hosted on DefinatelyTyped.
And i guess after all you are right. that might be the better way. I'm closing this PR

**Comment by jessetane** [comment]
Thanks a lot for looking into this further, I appreciate your effort and it's great to have learned a bit about TypeScript and had the chance to think about this particular problem. DefinatelyTyped seems like an interesting option - if you figure it out and want to open another PR to add a note / link in the readme to help other typescripters I'd be open to that.

**Comment by xogeny** [comment]
So as far as I can tell, the PR with local typings was never merged and nothing was ever submitted to DefinitelyTyped...is that correct?  It is really a shame since somebody went to all the trouble to develop the typings but the PR wasn't merged.  Personally, I prefer when projects include their own `.d.ts` files (vs. submitting to DefinitelyTyped).  This is especially true with mature libraries that are unlikely to have API changes. 😢

**Comment by jessetane** [comment]
@xogeny can you help me understand what's wrong with the DefinitelyTyped approach? Just so you can imagine it from my perspective: I don't currently use TS so adding this would put me in a position where I am unable to maintain parts of my own package.

**Comment by jessetane** [comment]
It seems like there is legitimate interest here so I don't want to shut this down if it's something the community wants..

**Comment by xogeny** [comment]
@jessetane So DefinitelyTyped is a giant git repository where TypeScript type definitions are stored.  The first problem is that it is a single repository.  Which means it is impossible to find anything specific to a given package among the issues, commits, PRs, etc. because there are so many things stored in there.  Second, the versioning isn't really supported and so when you install the type definitions from DefinitelyTyped, they don't necessarily correspond to the particular version of the package you are using.  Finally, using DefinitelyTyped means carrying around extra dependencies.  For every dependency `xyz`, you have an addition dependency on `@types/xyz`.

By keeping the type definitions in the repository of the main package, you only need to depend on a single library (*e.g.,* `queue`).  Furthermore, the type definitions are bundled with the NPM packages so you *always get the right version*.  If the library API changes, then the type definitions are updated as well and published right along side the Javascript so you never get any version skew.

So in my experience it is much simpler of the package itself stores the type definitions.  It is a total non-issue for the maintainer as long as the API never changes.  Said another way, you don't need to know TypeScript as long as somebody provides you with the type definitions and you don't change the API.  They just come along for the ride with each version.

**Comment by T4cC0re** [comment]
I can update the typings to match the current api (if there were any changes) and reopen the pull request. I never came around to adjust them for DefinitelyTyped, but if there is an interest now and this gets merged I will put in the effort.

**Comment by xogeny** [comment]
@T4cC0re If you are submitting type definitions to this repository, I don't see any point in also submitting them to DefinitelyTyped.  Is that what you were suggesting?  Or, am I missing something?

**Comment by jessetane** [comment]
I'm not sure why it's reasonable to assume the API won't change (it's changed quite a few times in the past), but if one or both of you is committed to updating the definitions before major releases, feel free to reopen.

Also, DefinitelyTyped isn't the only possible alternative, I haven't thought deeply about it but is the wrapper package approach definitely not of interest? Seems like if there was a `queue-ts` or similar, breaking changes could be released here without the need to update the type definitions synchronously..

**Comment by xogeny** [comment]
@jessetane I didn't mean to imply the API would never change.  I was just pointing out that it hasn't changed in at least 11 months as far as I can tell (last time the `README.md` was changed at least).  Since the `Queue` API is tied to the `Array` API, I also assume at least that part wouldn't change much.

All, I'm not a big fan of creating a `queue-ts` fork.  As I mentioned before, I'd rather see that definitions in this repository and published along side the code.  If that isn't acceptable, I'd rather just publish to DefinitelyTyped.  At least there you could annotate the types with some information indicating what version of `queue` they worked with.

Just to elaborate, the reason I don't like the `queue-ts` idea is that it is not a common pattern.  People won't know to look for it.  If I'm trying to use a package with TypeScript, the first thing I do is look for a `typings` field in `package.json`.  If that isn't there, I do a `yarn add -D @types/<package name>`.  If that fails, I just assume there are no types.  You could, of course, put something in the README about TypeScript typings being available in `queue-ts`, but I don't really see the point.  Using DefinitelyTyped as the same advantage (you can be out of sync with `queue`) but with the advantage that it is a "known" place to look.

In short, including the typings in this repo is the simplest for users.  The issue is having to support API changes.  If future versions of `queue` are backward compatible (a safe bet?), then the only risk is that the type definitions would be missing something, but they shouldn't "break" in that case.  Furthermore, any motivated user can submit a trivial PR to address any missing new functionality.  So this approach is easy, safe (for the reasons outlined in this paragraph) and, as a result of the safety, don't impose any real burden on the maintainer of the source, IMHO.

**Comment by T4cC0re** [comment]
@xogeny Good point. `npm i -S @types/<pkg>` is also my preferred way. You might be right, that people may not look for it.
I redact my previous comment. Really does not make much sense. If @jessetane shows will to merge, I will update this PR to fit the current API, if not, I will also update the typings but instead send a PR to definatelytyped. The latter option would be not that nice of course, but at least we would have typings

**Comment by jessetane** [comment]
Sounds like yall are up for becoming maintainers so, reopening :)

**Comment by Avinm** [comment]
Also it would be nice if the "queue" is exposed as a newable class instead so that one could use `const queue = new Queue(options)`

**Comment by Avinm** [comment]
+1
@jessetane, this pull request is nearly 2 years old, with TypeScript becoming increasingly popular, it'd be great if this could be merged soon.

**Comment by Avinm** [comment]
Is there something that needs to be done before this can be merged?

**Comment by jessetane** [comment]
yes there are merge conflicts

**Comment by T4cC0re** [comment]
Hi,

sorry for being absent. I had a lot on my plate.
@jessetane did you introduce breaking changes since I forked? If not, I would just fix the merge conflicts and typings could be used (altough not all features might be displayed).
I sadly do not have the time to update the typings to the latest state right now.

Could somebody else take over if this should require more work besides fixing merge conflicts?

**Comment by jessetane** [comment]
> did you introduce breaking changes since I forked?

I'm not sure... can someone check? I also don't know Typescript so it's a bit hard for me to verify the definitions are correct.

**Comment by jessetane** [comment]
If another person wants to open a new pr that's fine too, we can continue to discuss over there

**Comment by T4cC0re** [comment]
I will fix the merge conflicts really quick so someone can validate/take over from here. Last time I checked I just lack new features (such as promises) being exposed in the typings. Maybe my PR is a good starting point for someone else.

**Comment by jessetane** [comment]
oy the history got mucked up - first, the work should have been done in a branch (not the end of the world but still), and second instead of merging master, the patches should have been rebased - the way it is now makes the history very difficult to read.

**Comment by T4cC0re** [comment]
No Idea what Github just did there. I'll fix that via CLI

**Comment by jessetane** [comment]
you should be able to force push to your fork to fix things up

**Comment by T4cC0re** [comment]
Rebased it now.

**Comment by KeitIG** [comment]
I see the definitions are a bit outdated (no `autostart` or call signature on `queue`), I'll see if I can fix those
